src/Microsoft.OData.Core/DerivedTypeValidator.cs

```diff
@@ -0,0 +1,76 @@
+//---------------------------------------------------------------------
+// <copyright file="DerivedTypeValidator.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+namespace Microsoft.OData
+{
+    using System.Collections.Generic;
+    using System.Linq;
+    using Microsoft.OData.Edm;
+
+    /// <summary>
+    ///  Helper class to verify the resource type meets the drived type contraints.
+    /// </summary>
+    internal sealed class DerivedTypeValidator
+    {
+        private IEnumerable<string> derivedTypeConstraints;
+        private IEdmType expectedType;
+        private string resourceKind;
+        private string resourceName;
+
+        /// <summary>
+        /// Constructor.
+        /// </summary>
+        /// <param name="expectedType">The exptected type.</param>
+        /// <param name="derivedTypeConstraints">The derived type contraints.</param>
+        /// <param name="resourceKind">The resource type, be used at error message.</param>
+        /// <param name="resourceName">The resource name, be used at error message.</param>
+        public DerivedTypeValidator(IEdmType expectedType, IEnumerable<string> derivedTypeConstraints, string resourceKind, string resourceName)
+        {
+            this.derivedTypeConstraints = derivedTypeConstraints;
+            this.expectedType = expectedType;
+            this.resourceKind = resourceKind;
+            this.resourceName = resourceName;
+        }
+
+        /// <summary>
+        /// Validates the type of a resource.
+        /// </summary>
+        /// <param name="resourceType">The type of the resource.</param>
+        internal void ValidateResourceType(IEdmType resourceType)
+        {
+            if (resourceType == null)
+            {
+                return;
+            }
+
+            ValidateResourceType(resourceType.FullTypeName());
+        }
+
+        /// <summary>
+        /// Validates the full type name of a resource.
+        /// </summary>
+        /// <param name="resourceTypeName">The full type name of the resource.</param>
+        internal void ValidateResourceType(string resourceTypeName)
+        {
+            if (resourceTypeName == null)

```

Comment:
>resourceTypeName [](start = 16, length = 16)

Do we have cases where we would call this with a null resourceTypeName?

Reply:
At top level , ODL supports to create reader without the expected resource type. for example, see https://github.com/OData/odata.net/blob/master/src/Microsoft.OData.Core/ODataMessageReader.cs#L316-L319:

        /// <summary>Creates an <see cref="T:Microsoft.OData.ODataReader" /> to read a resource set.</summary>
        /// <returns>The created reader.</returns>
        public ODataReader CreateODataResourceSetReader()
        {
            return this.CreateODataResourceSetReader(/*entitySet*/null, /*expectedResourceType*/null);
        }

Reply:
:thumbsup: 

---
In reply to: [252894010](https://github.com/OData/odata.net/pull/1376#discussion_r252894010) [](ancestors = 252894010)